src/librustdoc/html/highlight.rs

```diff
@@ -440,12 +540,91 @@ fn exit_span(out: &mut Buffer) {
 ///     enter_span(Foo), string("text", None), exit_span()
 ///     string("text", Foo)
 /// ```
+///
 /// The latter can be thought of as a shorthand for the former, which is more
 /// flexible.
-fn string<T: Display>(out: &mut Buffer, text: T, klass: Option<Class>) {
+///
+/// Note that if `context` is not `None` and that the given `klass` contains a `Span`, the function
+/// will then try to find this `span` in the `span_correspondance_map`. If found, it'll then
+/// generate a link for this element (which corresponds to where its definition is located).
+fn string<T: Display>(
+    out: &mut Buffer,
+    text: T,
+    klass: Option<Class>,
+    context: Option<&Context<'_>>,
+    root_path: &str,
+) {
     match klass {
         None => write!(out, "{}", text),
-        Some(klass) => write!(out, "<span class=\"{}\">{}</span>", klass.as_html(), text),
+        Some(klass) => {
+            if let Some(def_span) = klass.get_span() {
+                let mut text = text.to_string();
+                if text.contains("::") {
+                    text =
+                        text.split("::").enumerate().fold(String::new(), |mut path, (pos, t)| {
+                            let pre = if pos != 0 { "::" } else { "" };
+                            match t {
+                                "self" | "Self" => write!(
+                                    &mut path,
+                                    "{}<span class=\"{}\">{}</span>",
+                                    pre,
+                                    Class::Self_((0, 0)).as_html(),
+                                    t
+                                ),
+                                "crate" | "super" => write!(
+                                    &mut path,
+                                    "{}<span class=\"{}\">{}</span>",
+                                    pre,
+                                    Class::KeyWord.as_html(),
+                                    t
+                                ),
+                                t => write!(&mut path, "{}{}", pre, t),
+                            }
+                            .expect("Failed to build source HTML path");
+                            path
+                        });
+                }
+                if let Some(context) = context {
+                    if let Some(href) =
```

Comment:
You can reduce this indent level:
```suggestion
                if let Some((context, href)) = context.and_then(|c| href.map(|h| (c, h)) {
```

Reply:
I originally did that but found it a bit hard to read but if you're ok with it, then so am I!

Reply:
So in the end it would give (because the `href` comes from the `span_map`'s span):

```rust

if let Some((context, span)) = context.and_then(|c| c.shared.span_correspondance_map.get(&def_span).map(|h| (c, h))) {
```

Reply:
I'll keep the code as is I think. ;)